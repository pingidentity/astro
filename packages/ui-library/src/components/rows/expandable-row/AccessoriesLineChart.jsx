import React from "react";
import PropTypes from "prop-types";
import { ResponsiveContainer, LineChart, XAxis, YAxis, Line, ReferenceLine } from "recharts";
import HelpHint from "../../tooltips/HelpHint";

import _ from "underscore";

/**
 * @class AccessoriesLineChart
 * @desc Component used only with expandable row component as an accessories.
 *
 * @param {string} [data-id="accessories-line-chart"]
 *     To define the base "data-id" value for the top-level HTML containers.
 * @param {string} [title]
 *     A string that serves as the title shown on the accessories.
 * @param {string} [count]
 *     A string that serves as the count shown on the accessories.
 * @param {string} [countLablel]
 *     A string that serves as the countLablel shown on the accessories.
 * @param {any} [hint]
 *     A string or JSX object that serves as the hint shown on hovered chart.
 * @param {any} [hintProps]
 *     A object that used for HelpHint as custom props.
 * @param {string} [chartLabel]
 *     A string that serves as the chartLablel shown below the chart.
 * @param {string} [trend]
 *     A string that serves as the chartLablel shown right from the chart.
 * @param {bool} [isTrendPositive=false]
 *     Whether the chart and trend color is blue or black.
 * @param {array} [data=[]]
 *     A list of objects that provides the data for the chart.
 * @param {string} [xAxisKey="id"]
 *    The object property that contains the x-axis data in the "data" prop
 * @param {string} [yAxisKey="value"]
 *    The object property that contains the y-axis data in the "data" prop
*  @param {function} [onLineChartClick]
 *     Click handler.
 *
 * @example
 *     <ExpandableRow.SimpleWrapper>
 *        <ExpandableRow
 *            title="Expanded Row with Line Chart"
 *            subtitle="Row Subtitle"
 *            expanded={false}
 *            onToggle={() => {}}
 *            rowAccessories={[
 *                <RowAccessoriesLineChart
 *                    key="row-accessories-line-chart"
 *                    title="Avg daily sign-ons:"
 *                    count="1,234,234"
 *                    countLabel="Past 7 days"
 *                    hint="See Contributing Data"
 *                    chartLabel="12 wk trend"
 *                    trend="+ 8.6%"
 *                    isTrendPositive={true}
 *                    data={[
 *                        { id: 1, value: 1 },
 *                        { id: 2, value: 5 },
 *                        { id: 3, value: 3 },
 *                        { id: 4, value: 2 },
 *                        { id: 5, value: 5 },
 *                        { id: 6, value: 1 },
 *                    ]}
 *                />,
 *                <Toggle key="toggle" />
 *            ]}
 *        />
 *     </ExpandableRow.SimpleWrapper>
 */

export default class AccessoriesLineChart extends React.Component {
    static displayName = "AccessoriesLineChart";

    static propTypes = {
        "data-id": PropTypes.string,
        title: PropTypes.string,
        count: PropTypes.string,
        countLabel: PropTypes.string,
        hint: PropTypes.any,
        hintProps: PropTypes.any,
        chartLabel: PropTypes.string,
        trend: PropTypes.string,
        isTrendPositive: PropTypes.bool,
        data: PropTypes.array.isRequired,
        xAxisKey: PropTypes.string,
        yAxisKey: PropTypes.string,
        onLineChartClick: PropTypes.func
    };

    static defaultProps = {
        "data-id": "accessories-line-chart",
        isTrendPositive: false,
        data: [],
        xAxisKey: "id",
        yAxisKey: "value",
    };

    _getTitle = () => {
        return this.props.title && <div key="title" className="accessories-line-chart__title">{this.props.title}</div>;
    };

    _getCount = () => {
        return this.props.count && (
            <div key="count" className="accessories-line-chart__count-wrapper">
                <div className="accessories-line-chart__count">{this.props.count}</div>
                <div className="accessories-line-chart__count-label">{this.props.countLabel}</div>
            </div>
        );
    };

    _getLineChart = () => {
        const trendColor = this.props.isTrendPositive ? "#2A60B1" : "#000";
        let referenceLinePoint;
        const isZeroTrend = _.chain(this.props.data)
            .filter(point => _.has(point, this.props.yAxisKey))
            .every(point => point[this.props.yAxisKey] === 0)
            .value();

        if (this.props.data[0] && !_.has(this.props.data[0], this.props.yAxisKey)) {
            referenceLinePoint = _.find(this.props.data.slice(1), point => _.has(point, this.props.yAxisKey)) ||
            _.last(this.props.data);
        }

        return (
            <div
                key="chart-wrapper"
                data-id={`${this.props["data-id"]}-chart-wrapper`}
                className="accessories-line-chart__chart-wrapper"
                onClick={this.props.onLineChartClick}
            >
                <div>
                    <div className="accessories-line-chart__chart">
                        {
                            _.isEmpty(this.props.data)
                                ? <div className="accessories-line-chart__no-data-chart" />
                                : <ResponsiveContainer width="100%" height="100%">
                                    <LineChart
                                        margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
                                        cursor="pointer"
                                        data={this.props.data}
                                    >
                                        <XAxis dataKey={this.props.xAxisKey} hide={true} />
                                        {
                                            isZeroTrend &&
                                           <YAxis domain={[-1, 1]} hide={true} />
                                        }
                                        <Line
                                            type="monotone"
                                            dot={false}
                                            isAnimationActive={false}
                                            dataKey={this.props.yAxisKey}
                                            stroke={trendColor}
                                            strokeWidth={1}
                                        />
                                        {
                                            referenceLinePoint && referenceLinePoint[this.props.xAxisKey] &&
                                            <ReferenceLine
                                                data-id="reference-line"
                                                x={referenceLinePoint[this.props.xAxisKey]}
                                                stroke="#686f77"
                                                strokeDasharray="2 2"
                                            />
                                        }
                                    </LineChart>
                                </ResponsiveContainer>

                        }

                    </div>
                    <div className="accessories-line-chart__chart-label">{this.props.chartLabel}</div>
                </div>
                <span className="accessories-line-chart__trend" style={{ color: trendColor }}>{this.props.trend}</span>
            </div>
        );
    };

    _getLineChartWithHint = () => {
        return (
            <div key="hint-wrapper" className="accessories-line-chart__hint-wrapper">
                <HelpHint
                    data-id={`${this.props["data-id"]}-helphint`}
                    placement="top"
                    hintText={this.props.hint}
                    {...this.props.hintProps}
                >
                    {this._getLineChart()}
                </HelpHint>
            </div>
        );
    };

    render() {
        return [
            this._getTitle(),
            this._getCount(),
            this.props.hint ? this._getLineChartWithHint() : this._getLineChart()
        ];
    }
}
