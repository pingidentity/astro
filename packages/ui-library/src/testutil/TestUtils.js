/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
