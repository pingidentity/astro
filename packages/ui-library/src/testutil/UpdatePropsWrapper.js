var React = require("react");
var assign = require("object-assign");

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
var UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = UpdatePropsWrapper;
