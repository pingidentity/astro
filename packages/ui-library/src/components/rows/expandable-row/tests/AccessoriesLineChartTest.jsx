window.__DEV__ = true;

import React from "react";
import { Line, ReferenceLine } from "recharts";
import { shallow } from "enzyme";
import _ from "underscore";
import { mountSnapshotDataIds } from "../../../../testutil/TestUtils";

import AccessoriesLineChart from "../AccessoriesLineChart";

describe("AccessoriesLineChartTest", function () {
    const defaultData = [
        { id: 1, value: "2" },
        { id: 2, value: "12" },
        { id: 3, value: "16" },
        { id: 4, value: "19" },
        { id: 5, value: "20" },
        { id: 6, value: "20" },
    ];

    const partialData = [
        { id: 1 },
        { id: 2 },
        { id: 3 },
        { id: 4, value: "19" },
        { id: 5, value: "20" },
        { id: 6, value: "20" },
    ];

    const emptyData = [
        { id: 1 },
        { id: 2 },
        { id: 3 },
        { id: 4 },
        { id: 5 },
        { id: 6 },
    ];

    const defaultProps = {
        key: "accessories-line-chart",
        title: "Avg daily sign-ons:",
        count: "1,234,234",
        countLabel: "Past 7 days",
        hint: "See Contributing Data",
        chartLabel: "12 wk trend",
        trend: "+ 8.6%",
        isTrendPositive: true,
        data: defaultData
    };

    function getComponent(customProps) {
        const props = _.defaults(customProps || {}, defaultProps);

        return shallow(<AccessoriesLineChart {...props} />);
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <AccessoriesLineChart
                hint="It's a secret"
                {...defaultProps}
            />
        );
    });

    it("should render with default data-id", function() {
        const component = getComponent();
        const accessories = component.getElements();

        expect(accessories.length).toBe(3);
    });

    it("should render without title", function() {
        const component = getComponent({ title: "" });
        const accessories = component.at(0).getElement();

        expect(accessories).toBe(null);
    });

    it("should render without count", function() {
        const component = getComponent({ count: "" });
        const accessories = component.at(1).getElement();

        expect(accessories).toBe(null);
    });

    it("should render without helphint", function() {
        const component = getComponent({ hint: "" });
        const сhartLineWrapper = component.at(2).getElement();

        expect(сhartLineWrapper.key).toBe("chart-wrapper");
    });

    it("should render chart and trend black if trand is negative", function() {
        const component = getComponent({ isTrendPositive: false });
        const chartLine = component.find(Line);

        expect(chartLine.prop("stroke")).toBe("#000");
    });

    it("should render without ReferenceLine if all data exist", function() {
        const component = getComponent();
        const isLineEmpty = component.find(ReferenceLine).isEmpty();

        expect(isLineEmpty).toBeTruthy();
    });

    it("should render ReferenceLine if not all data exist", function() {
        const component = getComponent({ data: partialData });
        const line = component.find(ReferenceLine).getElement();

        expect(line).toBeDefined();
    });

    it("should render ReferenceLine if have not data at all", function() {
        const component = getComponent({ data: emptyData });
        const line = component.find(ReferenceLine).getElement();

        expect(line).toBeDefined();
    });
});
