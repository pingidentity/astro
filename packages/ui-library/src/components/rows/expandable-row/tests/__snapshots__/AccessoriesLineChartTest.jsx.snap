// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessoriesLineChartTest data-id's don't change 1`] = `
Array [
  "accessories-line-chart-helphint",
  "accessories-line-chart-helphint-target",
  "accessories-line-chart-chart-wrapper",
  "tooltip",
]
`;
